### ISSUE TITLE
nn.Conv uses implicit rank promotion

### ISSUE THREAD

**Issue by khdlr** [patrick-kidger/equinox#308]
In the current implementation, `nn.Conv` uses implicit rank promotion. This causes errors when setting jax's `jax_numpy_rank_promotion` option to `'raise'`.

Minimal example:
```python
import jax
import equinox as eqx

conv = eqx.nn.Conv1d(16, 4, 1, key=jax.random.PRNGKey(0))
inputs = jax.numpy.ones([16, 8])

# Runs without error
conv(inputs)

# Raises a `Operands could not be broadcast together`
jax.config.update('jax_numpy_rank_promotion', 'raise')
conv(inputs)
```

Maybe I'm just paranoid, but I like setting the rank promotion behaviour to `'raise'` because I've spent way too much time debugging accidential rank promotions.

The source of the issue is that the bias has shape `(C, ...)`, but it is added to the convolution result with shape `(1, C, ...)` which has not been squeezed back to `(C, ...)` at that point. The straight-forward fix for this is probably to exchange the `squeeze` and addition of bias here: https://github.com/patrick-kidger/equinox/blob/fb098035d09fb18d2ecdaf44150977eb27701a48/equinox/nn/_conv.py#L191-L193

Happy to submit a PR for this if wanted.

**Comment by patrick-kidger** [comment]
Sounds good. I'd be very happy to take a PR on this!

**Comment by patrick-kidger** [comment]
Closing as this has been fixed for conv in #311 and for convtranspose in #335.
